src/read/pipelining.rs

```diff
@@ -0,0 +1,1209 @@
+        } else {
+            Ok(entry_path.split('/'))
+        }
+    }
+
+    /* TODO: consider using crate::unstable::path_to_string() for this--it involves new
+     * allocations, but that really shouldn't matter for our purposes. I like the idea of using our
+     * own logic here, since parallel/pipelined extraction is really a different use case than the
+     * rest of the zip crate, but it's definitely worth considering. */
+    pub(crate) fn normalize_parent_dirs<'a>(
+        entry_path: &'a str,
+    ) -> Result<(Vec<&'a str>, bool), PathSplitError> {
+        if entry_path.starts_with('/') || entry_path.starts_with('\\') {
+            return Err(PathSplitError::PathFormat(format!(
+                "path {:?} began with '/' or '\\' and is absolute",
+                entry_path
+            )));
+        }
+        let is_dir = is_dir(entry_path);
+
+        let mut ret: Vec<&'a str> = Vec::new();
+        for component in split_by_separator(entry_path)? {
+            match component {
+                /* Skip over repeated separators "//". We check separately for ending '/' with the
+                 * `is_dir` variable. */
+                "" => (),
+                /* Skip over redundant "." separators. */
+                "." => (),
+                /* If ".." is present, pop off the last element or return an error. */
+                ".." => {
+                    if ret.pop().is_none() {
+                        return Err(PathSplitError::PathFormat(format!(
+                        "path {:?} has too many '..' components and would escape the containing dir",
+                        entry_path
+                    )));
+                    }
+                }
+                _ => {
+                    ret.push(component);
+                }
+            }
+        }
+        if ret.is_empty() {
+            return Err(PathSplitError::PathFormat(format!(
+                "path {:?} resolves to the top-level directory",
+                entry_path
+            )));
+        }
+
+        Ok((ret, is_dir))
+    }
+
+    fn split_dir_file_components<'a, 's>(
+        all_components: &'s [&'a str],
+        is_dir: bool,
+    ) -> (&'s [&'a str], Option<&'a str>) {
+        if is_dir {
+            (all_components, None)
+        } else {
+            let (last, rest) = all_components.split_last().unwrap();
+            (rest, Some(last))
+        }
+    }
+
+    #[derive(PartialEq, Eq, Debug, Clone)]
+    pub(crate) struct DirEntry<'a, Data> {
+        pub properties: Option<Data>,
+        pub children: BTreeMap<&'a str, Box<FSEntry<'a, Data>>>,
+    }
+
+    impl<'a, Data> Default for DirEntry<'a, Data> {
+        fn default() -> Self {
+            Self {
+                properties: None,
+                children: BTreeMap::new(),
+            }
+        }
+    }
+
+    #[derive(PartialEq, Eq, Debug, Clone)]
+    pub(crate) enum FSEntry<'a, Data> {
+        Dir(DirEntry<'a, Data>),
+        File(Data),
```

Comment:
ZIP files can also contain symlinks.

Reply:
I wanted to avoid that at first (because it would require reading the symlink targets from the `ZipArchive`), but I think I'm ready to implement that pass now.